File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isUpper, isAlpha, isSpace)
import Data.Text as T

responseFor :: Text -> Text
responseFor prompt = T.pack response where
  response
    | silence = "Fine. Be that way!"
    | question && yelling = "Calm down, I know what I'm doing!"
    | question = "Sure."
    | yelling = "Whoa, chill out!"
    | otherwise = "Whatever."
  silence = T.null nospaces
  question = T.last nospaces == '?'
  yelling = allUpper (T.filter isAlpha nospaces)
  allUpper text = not (T.null text) &&  T.all isUpper text
  nospaces = T.filter (not . isSpace) prompt

File: nim/bob/bob.nim
import std/strutils
import std/sequtils

proc hey*(s: string): string =
  if isEmptyOrWhitespace(s):
    return "Fine. Be that way!"

  var is_question = s.strip().endswith("?")
  var letters = filter(s, isAlphaAscii)
  var is_yelling = len(letters) != 0 and len(filter(letters, isLowerAscii)) == 0

  if is_question and is_yelling:
    "Calm down, I know what I'm doing!"
  elif is_question:
    "Sure."
  elif is_yelling:
    "Whoa, chill out!"
  else:
    "Whatever."
    
    
